contracts/pair_astro_xastro/src/contract.rs

```diff
@@ -0,0 +1,232 @@
+};
+
+use astroport::asset::{Asset, AssetInfo};
+
+use astroport::pair::{ReverseSimulationResponse, SimulationResponse};
+use astroport::pair_bonded::{Config, ExecuteMsg};
+use astroport::querier::{query_supply, query_token_balance};
+use astroport::staking::Cw20HookMsg as StakingCw20HookMsg;
+use astroport_pair_bonded_template::base::PairBonded;
+use astroport_pair_bonded_template::error::ContractError;
+use astroport_pair_bonded_template::state::CONFIG;
+use cw20::Cw20ExecuteMsg;
+use cw_storage_plus::Item;
+
+/// This structure stores contract params.
+pub(crate) struct Contract<'a> {
+    pub params: Item<'a, Params>,
+}
+
+impl<'a> Contract<'a> {
+    pub(crate) fn new(params_key: &'a str) -> Self {
+        Contract {
+            params: Item::<Params>::new(params_key),
+        }
+    }
+}
+
+/// Implementation of the bonded pair template. Performs ASTRO-xASTRO swap operations.
+impl<'a> PairBonded<'a> for Contract<'a> {
+    fn swap(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        _info: MessageInfo,
+        sender: Addr,
+        offer_asset: Asset,
+        belief_price: Option<Decimal>,
+        max_spread: Option<Decimal>,
+        to: Option<Addr>,
+    ) -> Result<Response, ContractError> {
+        let config = CONFIG.load(deps.storage)?;
+
+        // If the asset balance already increased
+        // We should subtract the user deposit from the pool offer asset amount
+        let pools = config
+            .pair_info
+            .query_pools(&deps.querier, &env.contract.address)?
+            .into_iter()
+            .map(|mut p| {
+                if p.info.equal(&offer_asset.info) {
+                    p.amount = p.amount.checked_sub(offer_asset.amount)?;
+                }
+                Ok(p)
+            })
+            .collect::<StdResult<Vec<_>>>()?;
+
+        let offer_pool: Asset;
+        let ask_pool: Asset;
+
+        if offer_asset.info.equal(&pools[0].info) {
+            offer_pool = pools[0].clone();
+            ask_pool = pools[1].clone();
+        } else if offer_asset.info.equal(&pools[1].info) {
+            offer_pool = pools[1].clone();
+            ask_pool = pools[0].clone();
+        } else {
+            return Err(ContractError::AssetMismatch {});
+        }
+
+        let mut messages = vec![];
+
+        let params = self.params.load(deps.storage)?;
+
+        if offer_asset.info.equal(&AssetInfo::Token {
+            contract_addr: params.astro_addr.clone(),
+        }) {
+            messages.push(CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: params.astro_addr.to_string(),
+                msg: to_binary(&Cw20ExecuteMsg::Send {
+                    contract: params.staking_addr.to_string(),
+                    amount: offer_asset.amount,
+                    msg: to_binary(&StakingCw20HookMsg::Enter {})?,
+                })?,
+                funds: vec![],
+            }))
+        } else if offer_asset.info.equal(&AssetInfo::Token {
```

Comment:
This `if` is redundant as you already check that both assets belong to the pair in lines 64-72. I think you may just leave `else ...` here

Reply:
Ok, changed